src/System.Private.CoreLib/src/System/Diagnostics/Eventing/EventPipeEventDispatcher.cs

```diff
@@ -0,0 +1,162 @@
+            StopDispatchTask();
+
+            // Stop tracing.
+            // This is a no-op if it's already disabled.
+            EventPipeInternal.Disable();
+
+            // Check to see if tracing should be enabled.
+            if (m_subscriptions.Count <= 0)
+            {
+                return;
+            }
+
+            // Start collecting events.
+            EventKeywords aggregatedKeywords = EventKeywords.None;
+            EventLevel highestLevel = EventLevel.LogAlways;
+
+            foreach (EventListenerSubscription subscription in m_subscriptions.Values)
+            {
+                aggregatedKeywords |= subscription.MatchAnyKeywords;
+                highestLevel = (subscription.Level > highestLevel) ? subscription.Level : highestLevel;
+            }
+
+            EventPipeProviderConfiguration[] providerConfiguration = new EventPipeProviderConfiguration[]
+            {
+                new EventPipeProviderConfiguration(RuntimeEventSource.EventSourceName, (ulong) aggregatedKeywords, (uint) highestLevel)
+            };
+
+            EventPipeInternal.Enable(null, 1024, 1, providerConfiguration, 1);
+
+            // Start the dispatch task.
+            StartDispatchTask();
+        }
+
+        private void StartDispatchTask()
+        {
+            Debug.Assert(Monitor.IsEntered(m_dispatchControlLock));
+
+            if (m_dispatchTask == null)
+            {
+                m_stopDispatchTask = false;
+                m_dispatchTask = Task.Factory.StartNew(DispatchEventsToEventListeners, TaskCreationOptions.LongRunning);
+            }
+        }
+
+        private void StopDispatchTask()
+        {
+            Debug.Assert(Monitor.IsEntered(m_dispatchControlLock));
+
+            if(m_dispatchTask != null)
+            {
+                m_stopDispatchTask = true;
+                m_dispatchTask.Wait();
+                m_dispatchTask = null;
+            }
+        }
+
+        private unsafe void DispatchEventsToEventListeners()
+        {
+            // Struct to fill with the call to GetNextEvent.
+            EventPipeEventInstanceData instanceData;
+
+            while (!m_stopDispatchTask)
+            {
+                // Get the next event.
+                while (!m_stopDispatchTask && EventPipeInternal.GetNextEvent(&instanceData))
+                {
+                    // Filter based on provider.
+                    if (instanceData.ProviderID == m_RuntimeProviderID)
+                    {
+                        // Dispatch the event.
+                        ReadOnlySpan<Byte> payload = new ReadOnlySpan<byte>((void*)instanceData.Payload, (int)instanceData.PayloadLength);
+                        RuntimeEventSource.Log.ProcessEvent(instanceData.EventID, payload);
+                    }
+                }
+
+                // Wait for more events.
+                if (!m_stopDispatchTask)
+                {
+                    Thread.Sleep(10);
```

Comment:
We talked about this offline

This polling based mechanism is OK but here is an ugly issue on what this delay should be.   If the process is otherwise quiescent this keeps the CPU from going into a low power state.   It also introduces on average a 5 msec delay. 

We should consider the alternative of a AutoResetEvent where you pass back the Handle (it is a public API) to native code so they can signal it (Win32 SetEvent).  You can also signal it when you want to shut the thread down.

Reply:
Yup, and I have filed https://github.com/dotnet/coreclr/issues/18758.